File: surveying/benchmarks/lsm/Vector/LSM.hs

import qualified Data.Vector as DV
import Data.Vector (Vector)
import System.Random
import Data.Random.Normal

m, n_paths :: Int
m = 2            -- time steps
n_paths = 4 -- paths for valuation

r,vol,s0,t,v0_right,dt,df :: Double
r   = 0.06               -- short rate
vol = 0.2                -- volatility
s0  = 36.0               -- initial stock level
t   = 1.0                -- time - to - maturity
v0_right = 4.48637       -- American Put Option ( 500 steps bin . model)
dt  = t/(fromIntegral m) -- length of time interval
df  = exp(-r*dt)         -- discount factor per time interval

-- ^ Generate @n@ samples from a mean=0, stddev=1 normal distribution
rng :: Int -> IO (Vector Double)
rng n = do
  gen <- newStdGen
  return . DV.fromList . take n $ normals gen

-- ^ Monadic unfoldr. Stops unfolding after N steps and includes the
-- seed in the result.
unfoldrNM :: (Monad m) => (a -> m a) -> Int -> a -> m [a]
unfoldrNM _ 0 seed = return [seed]
unfoldrNM f n seed = do
  x <- f seed
  xs <- unfoldrNM f (n-1) x
  return (seed : xs)

-- ^ Generate paths
genPaths :: Int -> IO (Vector(Vector Double))
genPaths n = DV.fromList `fmap` unfoldrNM iter (m+1) initvec
  where
    initvec :: Vector Double
    initvec = DV.generate n (const s0)

    iter :: Vector Double -> IO (Vector Double)
    iter prev = do
      ran <- rng n
      return $ DV.zipWith (\s x -> s * exp((r-vol**2/2)*dt + vol*x*sqrt(dt))) prev ran

polyval :: Vector Double -> Double -> Double
polyval p x = DV.sum $ DV.zipWith (\p' n' -> p'*x**(fromIntegral n')) p pows
  where
   n = DV.length p
   pows = DV.fromList [n-1..0]

-- http://facstaff.unca.edu/mcmcclur/class/LinearII/presentations/html/leastsquares.html
vander :: Vector Double -> Vector (Vector Double)
vander xs = DV.generate (DV.length xs) (\x -> DV.map (** (fromIntegral x)) xs)

-- https://github.com/numpy/numpy/blob/master/numpy/lib/polynomial.py#L394
--polyfit :: 

main :: IO ()
main = do
  a <- genPaths 4
  print a

File: experiments/lsm/Python/LSM.py
# Valuation of American Put Option
# with Least-Squares Monte Carlo

from numpy import *
from numpy.random import standard_normal, seed
from time import time
t0 = time ()
## Simulation Parameters
seed(150000)       # seed for Python RNG
M   = 50           # time steps
I   = 4*4096       # paths for valuation
reg = 9            # no of basis functions
AP  = True         # antithetic paths
MM  = True         # moment matching of RN

## Parameters -- American Put Option                               
r   = 0.06         # short rate                                      
vol = 0.2          # volatility                                      
S0  = 36.0         # initial stock level                             
T   = 1.0          # time - to - maturity                            
V0_right = 4.48637 # American Put Option ( 500 steps bin . model )   
dt  = T/M          # length of time interval                         
df  = exp(-r*dt)   # discount factor per time interval


def RNG(I):
    """Generate I random numbers following specified AP and MM parameters"""
    if AP == True:
        ran = standard_normal(I/2)
        ran = concatenate((ran, -ran))
    else:
        ran = standard_normal(I)
    if MM == True:
        ran = ran-mean(ran)
        ran = ran/std(ran)
    return ran

def GenS(I):
    S = zeros((M+1, I), 'd')   # index level matrix
    S[0,:] = S0                # initial values
    for t in range(1, M+1, 1): # index level paths
        ran = RNG(I)
        S[t,:]= S[t-1,:] * exp((r-vol**2/2)*dt + vol*ran*sqrt(dt))
    return S

def IV(S):
    return maximum(40.0-S,0)

## Valuation by LSM
S = GenS(I) # generate stock price paths
h = IV(S)   # inner value matrix
V = IV(S)   # value matrix
for t in range(M-1, -1, -1):
    rg = polyfit(S[t,:], V[t+1,:]*df, reg)          # regression at time t
    C  = polyval(rg, S[t,:])                        # continuation values
    V[t,:] = where(h[t,:] > C, h[t,:], V[t+1,:]*df) # exercise decision
V0 = sum(V[0,:])/I # LSM estimator

## Output
print("Time elapsed in Seconds   %8.3f" % (time()-t0))
print("-----------------------------------------")
print("Right Value               %8.3f" % V0_right)
print("-----------------------------------------")
print("LSM Value for Am. Option  %8.3f" % V0)
print("Absolute Error            %8.3f" % (V0-V0_right))
print("Relative Error in Percent %8.3f" % ((V0-V0_right)/V0_right * 100))
print("-----------------------------------------")
